### PR TITLE
fixes #746 to support loading templates with number parameters

### LINKED ISSUE #746: Provide option to parse Templates without really parsing the underying objects using the typesafe DTO model
Issue:  When users specify replicas as a parameter in the templates 

replicas: '${REPLICAS}'
 
 and below line is executed
 
`List<Template> tlist = sourceClient().templates().inNamespace(namespace).withLabels(labelMap).list().getItems();`
 
it fails with an error 

```
caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Can not deserialize value of type java.lang.Integer from String "${REPLICAS}": not a valid Integer value
at [Source: N/A; line: -1, column: -1] (through reference chain: io.fabric8.openshift.api.model.TemplateList["items"]->java.util.ArrayList[0]->io.fabric8.openshift.api.model.Template["objects"]->java.util.ArrayList[0]->io.fabric8.openshift.api.model.DeploymentConfig["spec"]->io.fabric8.openshift.api.model.DeploymentConfigSpec["replicas"])

```

We are using 2.2.7 version

### PR DESCRIPTION
**Pull request by jstrachan**
now the processLocal(map) allows the numeric parameters to be replaced so that the template can be properly loaded providing the numeric properties are in the supplied map

### CODE REVIEW

File: openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateOperationsImpl.java [hunk 1]
```diff
@@ -214,4 +220,22 @@ private URL getProcessUrl() throws MalformedURLException {
     return new TemplateOperationsImpl(client, OpenShiftConfig.wrap(config), null, namespace, null, false, template, null, false, 0L,
       null, null, null, null, null);
   }
+
+  @Override
+  protected <T> T unmarshal(Class<T> type, InputStream is) throws IOException {
+    if (replaceValueStream != null) {
```
Comment (iocanel) [inline openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateOperationsImpl.java]: Apparently, I missing something here. `load()` is a method which I expect it to be called before `processLocally()` (which is the method that sets value to `replaceValueStream`) . So I don't see how its possible this not to be null.

File: openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateOperationsImpl.java [hunk 2]
```diff
@@ -139,6 +140,8 @@ public KubernetesList processLocally(ParameterValue... values) {
   }
 
   public KubernetesList processLocally(Map<String, String> valuesMap)  {
+    this.replaceValueStream = new ReplaceValueStream(valuesMap);
```
Comment (iocanel) [inline openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateOperationsImpl.java]: To answer my own question: `It's becase load() may call get() which may call unmarshal`.

File: openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateOperationsImpl.java (hunk 1 above)
Comment (iocanel) [inline openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/TemplateOperationsImpl.java]: Never mind, I figured it out :-)

**Comment by jstrachan** [comment]
@iocanel BTW my first attempt was to try custom deserialisers; turns out thats kinda complex as there are many numeric, date and boolean types in Java that all need changing - plus some Strings validation won't work with template expressions either. So a filter seems the simplest thing that could possibly work.

I hear you on the mutability thing though. Maybe we need a separate new method on TemplateOperations to `expandProperties(Map)` or something?

**Comment by jstrachan** [comment]
OK have added a `replaceParameters(Map)` method now that returns a new immutable `TemplateOperation` which is then internally reused by the `processLocally(Map)` method. It also means folks can load the template with parameter replacement too:
```
Template template = client.templates().withName("foo").replaceParameters(map).get();
```

**Comment by iocanel** [comment]
@jstrachan: I've implemented something like this on top of your pull request.

    client.templates().withParameters(myparms).list()
    client.templates().withParameters(myparams).withName('cheese').get()
    client.templates().withParameters(myparams).inNamespace('mynamespace').withName('cheese').get()

The `withParameters` is exclusive to template operations.

wdyt?

**Comment by jstrachan** [comment]
@iocanel looks awesome thanks!

**Comment by jstrachan** [comment]
@iocanel am happy to merge if you are?

**Comment by iocanel** [comment]
[merge]

**Pull request merged**